<a href="https://colab.research.google.com/github/gogowonji/Inhurdle_YOLOv4/blob/main/YOLOv4_Inhurdle.ipynb" target="_parent"><img src="https://colab.research.google.com/assets/colab-badge.svg" alt="Open In Colab"/></a>

# **1. 구글 드라이브 마운트하기**

In [ ]:
%cd ..
from google.colab import drive
drive.mount('/content/gdrive')

/
Mounted at /content/gdrive


mydrive로 심볼릭 링크 만들어줌

In [ ]:
!ln -s /content/gdrive/My\ Drive/ /mydrive

yolov4 폴더로 포커스

In [ ]:
%cd /mydrive/yolov4

/content/gdrive/My Drive/yolov4


#**2. Darknet clone**

In [ ]:
!git clone https://github.com/AlexeyAB/darknet

opencv, GPU 사용을 위해

In [ ]:
%cd darknet/
!sed -i 's/OPENCV=0/OPENCV=1/' Makefile
!sed -i 's/GPU=0/GPU=1/' Makefile
!sed -i 's/CUDNN=0/CUDNN=1/' Makefile
!sed -i 's/CUDNN_HALF=0/CUDNN_HALF=1/' Makefile
!sed -i 's/LIBSO=0/LIBSO=1/' Makefile

In [ ]:
!make

#**3. 학습을 위해 파일 옮겨주기**

In [ ]:
%cd data/
! find -maxdepth 1 -type f -exec rm -rf {} \;
'''데이터 지우는 파일 (시작)
%cd obj/
! find -maxdepth 1 -type f -exec rm -rf {} \;
%cd ..
데이터 지우는 파일 (끝)'''
%cd ..
%rm -rf cfg/
%mkdir cfg

/content/gdrive/MyDrive/yolov4/darknet/data
/content/gdrive/MyDrive/yolov4/darknet


In [ ]:
%cd /mydrive/yolov4

/content/gdrive/My Drive/yolov4


In [ ]:
!zip -FFv obj.zip --out obj2.zip
!unzip /mydrive/yolov4/obj2.zip -d data/

##데이터 총 파일 개수

In [ ]:
import os

folder_path = "/mydrive/yolov4/darknet/data/obj"
file_list = os.listdir(folder_path)
file_count = len(file_list)
print(file_count)

4672


## 수정된 파일 복사하여 yolov4 폴더에 넣기

이전 학습 시와 동일하다면 패스할 수 할 수 있다.










In [ ]:
!cp /mydrive/yolov4/yolov4-custom.cfg cfg

In [ ]:
!cp /mydrive/yolov4/obj.names data

In [ ]:
!cp /mydrive/yolov4/obj.data data

In [ ]:
!cp /mydrive/yolov4/process.py .

## 파이썬 파일 실행

dataset에서 trainset, testset 파일 나누어 주기

In [ ]:
!python process.py

/content/gdrive/My Drive/yolov4/darknet


In [ ]:
!pwd

/content/gdrive/My Drive/yolov4/darknet


#**4. 학습된 yolo4 weight 파일 다운로드**

In [ ]:
!wget https://github.com/AlexeyAB/darknet/releases/download/darknet_yolo_v3_optimal/yolov4.conv.137

--2023-06-07 12:05:39--  https://github.com/AlexeyAB/darknet/releases/download/darknet_yolo_v3_optimal/yolov4.conv.137
Resolving github.com (github.com)... 20.205.243.166
Connecting to github.com (github.com)|20.205.243.166|:443... connected.
HTTP request sent, awaiting response... 302 Found
Location: https://objects.githubusercontent.com/github-production-release-asset-2e65be/75388965/48bfe500-889d-11ea-819e-c4d182fcf0db?X-Amz-Algorithm=AWS4-HMAC-SHA256&X-Amz-Credential=AKIAIWNJYAX4CSVEH53A%2F20230607%2Fus-east-1%2Fs3%2Faws4_request&X-Amz-Date=20230607T120355Z&X-Amz-Expires=300&X-Amz-Signature=4946a546d28ecfafa68b38e5ec08f96b303118f91dbca0d1d29750a9667356f0&X-Amz-SignedHeaders=host&actor_id=0&key_id=0&repo_id=75388965&response-content-disposition=attachment%3B%20filename%3Dyolov4.conv.137&response-content-type=application%2Foctet-stream [following]
--2023-06-07 12:05:39--  https://objects.githubusercontent.com/github-production-release-asset-2e65be/75388965/48bfe500-889d-11ea-819e-c4d

#**5. 학습**

In [ ]:
from psutil import virtual_memory
ram_gb = virtual_memory().total / 1e9
print('Your runtime has {:.1f} gigabytes of available RAM\n'.format(ram_gb))

if ram_gb < 20:
  print('To enable a high-RAM runtime, select the Runtime > "Change runtime type"')
  print('menu, and then select High-RAM in the Runtime shape dropdown. Then, ')
  print('re-execute this cell.')
else:
  print('You are using a high-RAM runtime!')

In [ ]:
%cd cfg
!sed -i 's/batch=1/batch=64/' yolov4-custom.cfg
!sed -i 's/subdivisions=1/subdivisions=16/' yolov4-custom.cfg
%cd ..

/content/gdrive/MyDrive/yolov4/darknet/cfg
/content/gdrive/MyDrive/yolov4/darknet


RAM 용량 초과 - 학습 종료 문제로 크기 조절

In [ ]:
%cd cfg
!sed -i 's/width=512/width=256/' yolov4-custom.cfg
!sed -i 's/height=512/height=256/' yolov4-custom.cfg
%cd ..

/content/gdrive/MyDrive/yolov4/darknet/cfg
/content/gdrive/MyDrive/yolov4/darknet


처음 학습할 때

In [ ]:
!./darknet detector train data/obj.data cfg/yolov4-custom.cfg yolov4.conv.137 -dont_show -map

누적 학습할 때

In [ ]:
!./darknet detector train data/obj.data cfg/yolov4-custom.cfg backup/yolov4-custom_last.weights -dont_show -map

스트리밍 출력 내용이 길어서 마지막 5000줄이 삭제되었습니다.
v3 (iou loss, Normalizer: (iou: 0.07, obj: 1.00, cls: 1.00) Region 161 Avg (IOU: 0.846074), count: 4, class_loss = 0.002818, iou_loss = 0.853406, total_loss = 0.856224 
 total_bbox = 673182, rewritten_bbox = 0.166523 % 
v3 (iou loss, Normalizer: (iou: 0.07, obj: 1.00, cls: 1.00) Region 139 Avg (IOU: 0.740877), count: 5, class_loss = 1.040928, iou_loss = 31.955860, total_loss = 32.996788 
v3 (iou loss, Normalizer: (iou: 0.07, obj: 1.00, cls: 1.00) Region 150 Avg (IOU: 0.737796), count: 8, class_loss = 4.779710, iou_loss = 6.925252, total_loss = 11.704962 
v3 (iou loss, Normalizer: (iou: 0.07, obj: 1.00, cls: 1.00) Region 161 Avg (IOU: 0.881736), count: 5, class_loss = 1.899355, iou_loss = 1.770533, total_loss = 3.669888 
 total_bbox = 673200, rewritten_bbox = 0.166518 % 
v3 (iou loss, Normalizer: (iou: 0.07, obj: 1.00, cls: 1.00) Region 139 Avg (IOU: 0.901579), count: 6, class_loss = 0.033914, iou_loss = 34.166565, total_loss = 34.200478 
v3 (iou loss

### 학습 완료 결과

Tensor Cores are used.
 Last accuracy mAP@0.50 = 89.24 %, best = 90.50 %
 8000: 1.281314, 0.924117 avg loss, 0.000010 rate, 10.949139 seconds, 1328000 images, 0.999079 hours left
Resizing to initial size: 256 x 256  try to allocate additional workspace_size = 135.27 MB
 CUDA allocate done!

 calculation mAP (mean average precision)...
 Detection layer: 139 - type = 28
 Detection layer: 150 - type = 28
 Detection layer: 161 - type = 28
228
 detections_count = 1511, unique_truth_count = 520  
class_id = 0, name = bollard, ap = 94.77%   	 (TP = 243, FP = 16)
class_id = 1, name = pole, ap = 80.39%   	 (TP = 87, FP = 13)
class_id = 2, name = person, ap = 94.11%   	 (TP = 46, FP = 7)
class_id = 3, name = etc, ap = 91.07%   	 (TP = 79, FP = 8)

 for conf_thresh = 0.25, precision = 0.91, recall = 0.88, F1-score = 0.89
 for conf_thresh = 0.25, TP = 455, FP = 44, FN = 65, average IoU = 76.56 %

 IoU threshold = 50 %, used Area-Under-Curve for each unique Recall
 mean average precision (mAP@0.50) = 0.900853, or 90.09 %
Total Detection Time: 49 Seconds

Set -points flag:
 `-points 101` for MS COCO
 `-points 11` for PascalVOC 2007 (uncomment `difficult` in voc.data)
 `-points 0` (AUC) for ImageNet, PascalVOC 2010-2012, your custom dataset

 mean_average_precision (mAP@0.50) = 0.900853
Saving weights to /mydrive/yolov4/darknet/backup/yolov4-custom_8000.weights
Saving weights to /mydrive/yolov4/darknet/backup/yolov4-custom_last.weights
Saving weights to /mydrive/yolov4/darknet/backup/yolov4-custom_final.weights
If you want to train from the beginning, then use flag in the end of training command: -clear

#**6. Test전 준비단계**

**imshow() 구현**

In [ ]:
# function imShow define
def imShow(path):
  import cv2
  import matplotlib.pyplot as plt
  %matplotlib inline

  image = cv2.imread(path)
  height, width = image.shape[:2]
  resized_image = cv2.resize(image,(3*width, 3*height), interpolation = cv2.INTER_CUBIC)

  fig = plt.gcf()
  fig.set_size_inches(18, 10)
  plt.axis("off")
  plt.imshow(cv2.cvtColor(resized_image, cv2.COLOR_BGR2RGB))
  plt.show()

cfg파일에서 batch, subdivisions 1로 바꾸기

In [ ]:
%cd cfg
!sed -i 's/batch=64/batch=1/' yolov4-custom.cfg
!sed -i 's/subdivisions=16/subdivisions=1/' yolov4-custom.cfg
%cd ..

/content/gdrive/MyDrive/yolov4/darknet/cfg
/content/gdrive/MyDrive/yolov4/darknet


# **7. TEST**

## **a. 기둥**

test 폴더에 있는 이미지 파일로 테스트

In [ ]:
!./darknet detector test data/obj.data cfg/yolov4-custom.cfg /mydrive/yolov4/darknet/backup/yolov4-custom_best.weights /mydrive/test/image2.jpg -thresh 0.3
imShow('predictions.jpg')

In [ ]:
!./darknet detector test data/obj.data cfg/yolov4-custom.cfg /mydrive/yolov4/darknet/backup/yolov4-custom_best.weights /mydrive/test/image3.jpg -thresh 0.3
imShow('predictions.jpg')

In [ ]:
!./darknet detector test data/obj.data cfg/yolov4-custom.cfg /mydrive/yolov4/darknet/backup/yolov4-custom_best.weights /mydrive/test/image5.jpg -thresh 0.3
imShow('predictions.jpg')

In [ ]:
!./darknet detector test data/obj.data cfg/yolov4-custom.cfg /mydrive/yolov4/darknet/backup/yolov4-custom_best.weights /mydrive/test/pole.jpg -thresh 0.3
imShow('predictions.jpg')

## **b. 볼라드 기둥 etc**

In [ ]:
!./darknet detector test data/obj.data cfg/yolov4-custom.cfg /mydrive/yolov4/darknet/backup/yolov4-custom_best.weights /mydrive/test/image6.jpg -thresh 0.3
imShow('predictions.jpg')

In [ ]:
!./darknet detector test data/obj.data cfg/yolov4-custom.cfg /mydrive/yolov4/darknet/backup/yolov4-custom_best.weights /mydrive/test/outdoor2.jpg -thresh 0.3
imShow('predictions.jpg')

In [ ]:
!./darknet detector test data/obj.data cfg/yolov4-custom.cfg /mydrive/yolov4/darknet/backup/yolov4-custom_best.weights /mydrive/test/outdoor3.jpg -thresh 0.3
imShow('predictions.jpg')

In [ ]:
!./darknet detector test data/obj.data cfg/yolov4-custom.cfg /mydrive/yolov4/darknet/backup/yolov4-custom_best.weights /mydrive/test/bollard1.jpg -thresh 0.3
imShow('predictions.jpg')

In [ ]:
!./darknet detector test data/obj.data cfg/yolov4-custom.cfg /mydrive/yolov4/darknet/backup/yolov4-custom_best.weights /mydrive/test/etc1.jpg -thresh 0.3
imShow('predictions.jpg')

In [ ]:
!./darknet detector test data/obj.data cfg/yolov4-custom.cfg /mydrive/yolov4/darknet/backup/yolov4-custom_best.weights /mydrive/test/etc2.jpg -thresh 0.3
imShow('predictions.jpg')

##**c. 사람**

In [ ]:
!./darknet detector test data/obj.data cfg/yolov4-custom.cfg /mydrive/yolov4/darknet/backup/yolov4-custom_best.weights /mydrive/test/people.jpg -thresh 0.3
imShow('predictions.jpg')

In [ ]:
!./darknet detector test data/obj.data cfg/yolov4-custom.cfg /mydrive/yolov4/darknet/backup/yolov4-custom_best.weights /mydrive/test/people1.jpg -thresh 0.3
imShow('predictions.jpg')

In [ ]:
!./darknet detector test data/obj.data cfg/yolov4-custom.cfg /mydrive/yolov4/darknet/backup/yolov4-custom_best.weights /mydrive/test/people2.jpg -thresh 0.3
imShow('predictions.jpg')

In [ ]:
!./darknet detector test data/obj.data cfg/yolov4-custom.cfg /mydrive/yolov4/darknet/backup/yolov4-custom_last.weights /mydrive/test/people3.jpg -thresh 0.3
imShow('predictions.jpg')

In [ ]:
!./darknet detector test data/obj.data cfg/yolov4-custom.cfg /mydrive/yolov4/darknet/backup/yolov4-custom_last.weights /mydrive/test/people4.jpg -thresh 0.3
imShow('predictions.jpg')

## **d. 추가 test 파일**

In [ ]:
!./darknet detector test data/obj.data cfg/yolov4-custom.cfg /mydrive/yolov4/darknet/backup/yolov4-custom_best.weights /mydrive/test/1.jpg -thresh 0.3
imShow('predictions.jpg')

In [ ]:
!./darknet detector test data/obj.data cfg/yolov4-custom.cfg /mydrive/yolov4/darknet/backup/yolov4-custom_best.weights /mydrive/test/2.jpg -thresh 0.3
imShow('predictions.jpg')

In [ ]:
!./darknet detector test data/obj.data cfg/yolov4-custom.cfg /mydrive/yolov4/darknet/backup/yolov4-custom_best.weights /mydrive/test/3.jpg -thresh 0.3
imShow('predictions.jpg')

In [ ]:
!./darknet detector test data/obj.data cfg/yolov4-custom.cfg /mydrive/yolov4/darknet/backup/yolov4-custom_best.weights /mydrive/test/4.jpg -thresh 0.3
imShow('predictions.jpg')

In [ ]:
!./darknet detector test data/obj.data cfg/yolov4-custom.cfg /mydrive/yolov4/darknet/backup/yolov4-custom_best.weights /mydrive/test/5.jpg -thresh 0.3
imShow('predictions.jpg')

In [ ]:
!./darknet detector test data/obj.data cfg/yolov4-custom.cfg /mydrive/yolov4/darknet/backup/yolov4-custom_best.weights /mydrive/test/6.jpg -thresh 0.3
imShow('predictions.jpg')

In [ ]:
!./darknet detector test data/obj.data cfg/yolov4-custom.cfg /mydrive/yolov4/darknet/backup/yolov4-custom_best.weights /mydrive/test/7.jpg -thresh 0.3
imShow('predictions.jpg')

In [ ]:
!./darknet detector test data/obj.data cfg/yolov4-custom.cfg /mydrive/yolov4/darknet/backup/yolov4-custom_best.weights /mydrive/test/8.jpg -thresh 0.3
imShow('predictions.jpg')

##**e. 비디오**

In [ ]:
!./darknet detector demo data/obj.data cfg/yolov4-custom.cfg /mydrive/yolov4/darknet/backup/yolov4-custom_best.weights -dont_show /mydrive/test/video1.mov -thresh 0.3 -i 0 -out_filename /mydrive/test/results.avi

In [ ]:
!./darknet detector demo data/obj.data cfg/yolov4-custom.cfg /mydrive/yolov4/darknet/backup/yolov4-custom_best.weights -dont_show /mydrive/test/video2.mov -thresh 0.3 -i 0 -out_filename /mydrive/test/results2.avi

In [ ]:
!./darknet detector demo data/obj.data cfg/yolov4-custom.cfg /mydrive/yolov4/darknet/backup/yolov4-custom_best.weights -dont_show /mydrive/test/video3.mov -thresh 0.3 -i 0 -out_filename /mydrive/test/results3.avi

In [ ]:
!./darknet detector demo data/obj.data cfg/yolov4-custom.cfg /mydrive/yolov4/darknet/backup/yolov4-custom_best.weights -dont_show /mydrive/test/video6.mov -thresh 0.3 -i 0 -out_filename /mydrive/test/results6.avi